sa/sa.go

```diff
@@ -1075,6 +1087,60 @@ func addFQDNSet(tx *gorp.Transaction, names []string, serial string, issued time
 	})
 }
 
+// addOrderFQDNSet creates a new OrderFQDNSet row using the provided
+// information. This function accepts a transaction so that the orderFqdnSet
+// addition can take place within the order addition transaction.
+func addOrderFQDNSet(
+	tx *gorp.Transaction,
+	names []string,
+	orderID int64,
+	regID int64,
+	expires time.Time) error {
+	if err := tx.Insert(&orderFQDNSet{
+		SetHash:        hashNames(names),
+		OrderID:        orderID,
+		RegistrationID: regID,
+		Expires:        expires,
+	}); err != nil {
+		return Rollback(tx, err)
+	}
+	return nil
+}
+
+// deleteOrderFQDNSet deletes a OrderFQDNSet row that matches the provided
+// information. This function accepts a transaction so that the deletion can
+// take place within the finalization transaction.
+func deleteOrderFQDNSet(
+	tx *gorp.Transaction,
+	names []string,
+	orderID int64,
+	regID int64) error {
+
+	result, err := tx.Exec(`
+	  DELETE FROM orderFqdnSets
```

Comment:
This query has a critical reliance on expired `orderFqdnSets` being purged promptly. There's no index on orderID, so the query would use the setHash index, then scan for orderID. This works fine in the common case, but if there are lots of expired orders for the same setHash, it would be slow. I think we should solve this by adding an orderID index on orderFQDNSet. Once that's done this query (and function call) can be simplified to just take an orderID, no regID or names.

Reply:
Resolved in 4038e64